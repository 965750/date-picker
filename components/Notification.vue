<template>
  <div v-if="error" class="notification">
    <p>{{ error }}</p>
  </div>
</template>

<script>
export default {
  name: 'Notification',
  props: {
    pickedDates: {
      required: true,
      type: Object
    },
    availableDates: {
      required: true,
      type: Array
    }
  },
  data () {
    return {
      error: null
    }
  },
  watch: {
    pickedDates: {
      handler (val) {
        if (val.start && val.end) {
          this.validate()
        }
      },
      deep: true
    }
  },
  methods: {
    validate () {
      const startTimestamp = this.pickedDates.start.timestamp
      const endTimestamp = this.pickedDates.end.timestamp

      let error = null
      let rangeCounter = this.availableDates.length

      this.availableDates.forEach((date) => {
        if (
          startTimestamp <= date.end.getTime() &&
          startTimestamp >= date.start.getTime() &&
          endTimestamp <= date.end.getTime() &&
          endTimestamp >= date.start.getTime()
        ) {
          error = null
          rangeCounter--
        }
      })

      if (rangeCounter === this.availableDates.length) {
        error = 'You should pick date from range of available'
      }

      if (startTimestamp > endTimestamp) {
        error = 'Check In date should be earlier than Check Out date'
      }

      this.error = error
    }
  }
}
</script>

<style lang="scss">
@import '../assets/css/variables';

.notification {
  background: $error;
  border-radius: 4px;
  padding: 10px;
  color: #FFF;
  margin-top: 10px;
  text-align: center;
}
</style>
