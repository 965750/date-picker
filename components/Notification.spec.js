import {
  shallowMount,
  createLocalVue
} from '@vue/test-utils'
import Notification from './Notification'

const localVue = createLocalVue()

const initialProps = {
  pickedDates: {
    start: {
      label: 4,
      timestamp: 1567548000000
    },
    end: {
      label: 9,
      timestamp: 1567980000000
    }
  },
  availableDates: [
    {
      start: new Date(2019, 8, 3),
      end: new Date(2019, 8, 10)
    },
    {
      start: new Date(2019, 8, 22),
      end: new Date(2019, 8, 27)
    },
    {
      start: new Date(2019, 9, 8),
      end: new Date(2019, 9, 24)
    }
  ]
}

describe('Notification.vue', () => {
  const createWrapper = (props = initialProps) => shallowMount(Notification, {
    localVue,
    propsData: {
      ...props
    }
  })

  it('should match empty snapshot when dates were picked correctly', () => {
    const wrapper = createWrapper()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should activate error notification when dates were picked out of the available range', () => {
    const wrapper = createWrapper()

    wrapper.setProps({
      pickedDates: {
        start: {
          label: 1,
          timestamp: 1567288800000
        },
        end: {
          label: 9,
          timestamp: 1567980000000
        }
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should activate error notification when check out date is earlier than check in date', () => {
    const wrapper = createWrapper()

    wrapper.setProps({
      pickedDates: {
        start: {
          label: 9,
          timestamp: 1567980000000
        },
        end: {
          label: 4,
          timestamp: 1567548000000
        }
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should NOT activate error notification when only one date was picked', () => {
    const wrapper = createWrapper()

    wrapper.setProps({
      pickedDates: {
        start: {
          label: 9,
          timestamp: 1567980000000
        },
        end: null
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
