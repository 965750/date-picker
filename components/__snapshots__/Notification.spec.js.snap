// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification.vue should NOT activate error notification when only one date was picked 1`] = `<!---->`;

exports[`Notification.vue should activate error notification when check out date is earlier than check in date 1`] = `
<div
  class="notification"
>
  <p>
    Check In date should be earlier than Check Out date
  </p>
</div>
`;

exports[`Notification.vue should activate error notification when dates were picked out of the available range 1`] = `
<div
  class="notification"
>
  <p>
    You should pick date from range of available
  </p>
</div>
`;

exports[`Notification.vue should match empty snapshot when dates were picked correctly 1`] = `<!---->`;
